### PR TITLE
provide core events for started and finished commands

### PR DESCRIPTION
**Pull request by flavius**
The purpose of the new events is to make it easy for users to detect executed commands, before and after the fact, as they are seen by the OS, after all the xonsh magic has been done.

It's just a working example. There are at least two rough edges I am aware of:

* the name of the two events is confusing; we need better names
* the "before" event just passes over the internal representation of the command - the purpose of these events is to offer a simple hooking point for basic needs, so I would prefer a quick and dirty access to the command(s) as lists of strings with the commands and their parameters

I have experience with formal languages, but my idiomatic python is rusty (pun intended) and I've never written a shell, so any advice is welcome.

### CODE REVIEW

**Comment by flavius** [comment]
@astronouth7303 I've noticed the pre/post convention, I just don't know how to name this pair of events. Once I have a name, I'll add it to news and to docs. I've been thinking hard, so any suggestion is welcome.

File: xonsh/built_ins.py [hunk 1]
```diff
@@ -792,6 +792,7 @@ def run_subproc(cmds, captured=False):
     Lastly, the captured argument affects only the last real command.
     """
     specs = cmds_to_specs(cmds, captured=captured)
+    events.on_command_started.fire(cmd=specs)
```
Comment (scopatz) [inline xonsh/built_ins.py]: The keyword here should be `specs`, not `cmd`.  Also this isn't really after the command started. It is more like "after commands specified" or "before pipeline execution"

File: xonsh/proc.py [hunk 2]
```diff
@@ -649,6 +652,7 @@ def run(self):
         # kill the process if it is still alive. Happens when piping.
         if proc.poll() is None:
             proc.terminate()
+        events.on_command_finished.fire(cmd=self.command_args, sig=self.previous_signal)
```
Comment (scopatz) [inline xonsh/proc.py]: I don't think this should be on PopenThread.  It makes the most sense for me for this to be in `CommandPipeline.end()`

**Comment by scopatz** [comment]
Thanks for putting this in @flavius!

**Comment by astronouth7303** [comment]
> There are only two hard things in Computer Science: cache invalidation and naming things. -- Phil Karlton

I would suggest `on_pre_process`, but that sounds a lot like the function of `on_transform_command`.

Maybe `on_pre_execution` or `on_pre_exec`?

@scopatz from a design and API usability standpoint, would it make more sense to do this the moment before or the moment after processes are executed?

**Comment by flavius** [comment]
Well, the decision whether to trigger the pre event before or after opening the process depends on the semantics.

* a pre event before the execution was started is more like an intent. It has its own right to exist
* a pre event after the execution was started, is a (post) fact

Both of them could have their usages. Case 1 could prepare a special environment in some way for the command, case 2 could be blocked by the command to do that, in case of locked resources for instance.

For the sake of symmetry with the post event, I prefer Case 1, which can only fail when the command does not exist, which would be easy to check by the hook itself, if it needs to.

**Comment by scopatz** [comment]
> @scopatz from a design and API usability standpoint, would it make more sense to do this the moment before or the moment after processes are executed?

@astronouth7303 I don't think so. I think it would make it more confusing.  Because we have 4 different ways of running commands, it would be unnecessarily duplicated code.  Also do they each get their own event or different events?  Also what happens if someone aliases `git` form a subprocess to some internal command: which event picks that up? How does the event handler know which element of a pipeline a command is if it only gets the single command?  

I think that those are some of the hard questions that would need to be answered if we went with per command events instead of per pipeline events.  Add to that that fact that the SubprocSpec and CommandPipeline are the only "public" parts of the API. That is, I have been working off the model that once a CommandPipeline has started it is in complete control and the only side effects are job control related.

In short, it is much easier to reason about pipelines than commands. I think that is what we should expose to the user too. At least at a first pass, since we'll always need to provide a way to get the pipeline.

**Comment by scopatz** [comment]
@flavius - Just wanna make sure that this doesn't get lost.  Let us know when you think it is ready for another look.

**Comment by flavius** [comment]
@scopatz Sorry for the silence, I got caught in some work. I'll be back to this quite soon and I'll update this PR.

**Comment by gforsyth** [comment]
Hi @flavius -- I'm going through and closing all pull requests that have been idle for more than a year.  We'd love to have your contribution and if you come back to it, please feel free to either open a new PR or you can ping here to have us reopen the existing PR.

Thanks!
